src/charm.py

```diff
@@ -129,6 +129,9 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:
         # check whether restart is needed for all `*_changed` events
         self.on[self.restart.name].acquire_lock.emit()
 
+        if self.tls.upgrading and len(self.cluster.peer_units) == 1:
```

Comment:
Just a questions: is this line needed to update the quorum to SSL later in the next event run (on L123) after the upgrade is complete?

Reply:
Config settings are applied to the unit during `_restart()`.
During `_restart()`, WHAT config actually gets set is dependent on the relation data currently active.

For this, we need to do 1 run of L123`update_quorum`, it sets some data, unit restarts.
If we have multiple app units, other units then get `relation_changed`, do something, causing this unit to run again, this time now setting some new data, and restarting.
For single unit apps, they don't get their own `relation_changed`, so it can't complete the 2 necessary `relation_changed` runs and don't fully upgrade to SSL.

We could probably have added a specific method for this edge-case, but that seemed bloaty, hence the defer.

Did that make sense?

Reply:
It makes perfect sense Marc. I was thinking about that flow, but had not all the details. Thanks for the in depth explanation!